pilot/pkg/proxy/envoy/watcher.go

```diff
@@ -48,24 +48,41 @@ type CertSource struct {
 }
 
 type watcher struct {
-	certs   []CertSource
-	updates chan<- interface{}
+	certs        []CertSource
+	updates      chan<- interface{}
+	waitForCerts bool
 }
 
 // NewWatcher creates a new watcher instance from a proxy agent and a set of monitored certificate paths
 // (directories with files in them)
-func NewWatcher(
-	certs []CertSource,
-	updates chan<- interface{}) Watcher {
+func NewWatcher(certs []CertSource, updates chan<- interface{}, waitForCerts bool) Watcher {
 	return &watcher{
-		certs:   certs,
-		updates: updates,
+		certs:        certs,
+		updates:      updates,
+		waitForCerts: waitForCerts,
 	}
 }
 
 func (w *watcher) Run(ctx context.Context) {
-	// kick start the proxy with partial state (in case there are no notifications coming)
-	w.SendConfig()
+	shouldKickStart := true
+	if w.waitForCerts {
+		// If none of the cert/key files exist, should not kick start.
+		shouldKickStart = false
+		for _, cert := range w.certs {
+			for _, filename := range cert.Files {
```

Comment:
This loop is busy waiting, right?
Maybe add a short delay between the iterations?

Reply:
Because now sending config is not guaranteed we need to make sure that there is no race between watching files and their creation.
We want to make sure that files where not created before starting watch. 
To do that we firstly watch all files and in same goroutine we check if any any file exists.

Reply:
@ymesika This loop doesn't wait. It's only to check if the cert files exist after the Pilot agent starts. If the waitForCerts is specified and the files don't exist, the watcher will be the one to start Envoy once the files exist. Otherwise, Envoy is always started when Pilot agent starts, regardless of the cert files.

Reply:
@serathius That makes sense, I will move the watcher upfront.

Reply:
> This loop doesn't wait.

@myidpt this is what I'm worried about when I asked whether it's [busy waiting loop](https://en.wikipedia.org/wiki/Busy_waiting). The loop keeps executing the `os.*` functions until they exist instead having a (short) wait between iterations.